### DISCUSSION TITLE
Allow asset downloads/transformations through JS SDK

### DISCUSSION THREAD

**Discussion by ametis70** [directus/directus#9201] [Feature Requests]
Hello, I was reading the docs and realized that [file transforms can be requested through the `/assets/<id>?<parameters>`](https://docs.directus.io/reference/files/#requesting-a-thumbnail), but I believe that functionality is not available in the JS-SDK yet.

I could try to add this functionality and submit a pull request if that's ok.
Sorry in advance if I missed something!

**Comment by rijkvanzanten** [comment]
Heya! That endpoint will return the actual file data. Is that what you're looking for in the SDK? As of right now, you could already add that endpoint straight into a `<img>` tag and have it work 🙂

**Reply by ametis70** [reply]
Thanks for the quick response. Yes, from what I've tried, a file uploaded to Directus can be requested with `/assets/<id>`, and that same file can be requested using parameters to perform some transformation using sharp. For instance,
```
/assets/1?width=500&height=500&format=webp
```
will return a `.webp` of `500x500px` even if the file that was uploaded didn't have that dimensions or format.

Using the JS-SDK, that same file can be requested using:
```js
const client = new Directus()
await client.auth.login()
const file = await client.files.readOne(1)
```
But I couldn't find a way to transform the requested image inside js. I'm thinking of something in the lines of:

```js
const file = await client.files.readOne(1, { 
  transform: {
    width: 500,
    height: 500,
    format: 'webp'
  }
})
```

This would have the niceties of writing JS/TS (instead of a long string with the parameters, that is more error prone). This would also allow to write some serverside logic like a caching middleware (not related to Directus), to prevent generating the same image multiple times

**Reply by rijkvanzanten** [reply]
Yeah! Im just curious what you'd use that "const file" for after the fact, and what it would hold 🤔

**Reply by ametis70** [reply]
Oh, right. I didn't think of that :man_facepalming:

To give some context, right now I'm using Directus with a static site generator (Next.js), and I'm using sharp transforms using the API endpoint to download a few images in multiple sizes and saving them to the filesystem with Node.js.

Here is a somewhat simplified snippet with two functions:

1. `downloadFile` returns a promise that will resolve to undefined and has the side effect of downloading a file. 
2. `downloadImage` will check if the file exists, and if it doesn't it will generate the query parameters string and call `downloadFile`.

```ts
async function downloadFile(
  resource: string,
  downloadPath: string,
): Promise<void> {
  const bearer = 'Bearer ' + client.auth.token
  const url = `${process.env.PUBLIC_URL}${resource}`
  const response = await fetch(url, { headers: { Authorization: bearer } })
  const arrayBuffer = await response.arrayBuffer()
  const buffer = Buffer.from(arrayBuffer)

  fs.createWriteStream(downloadPath).write(buffer)
  console.info(`Downloaded ${downloadPath}`)
}

type ImageTransform = {
  fit?: 'cover' | 'contain' | 'inside' | 'outside'
  width: number
  height?: number
  quality?: number
  withoutEnlargement?: boolean
  format: 'jpg' | 'png' | 'webp' | 'tiff'
  transforms?: unknown[]
}

async function downloadImage (id, directory, params: ImageTransform) {
  const filename = `${id}-${params.width}w.${params.format}`

  const filePath = path.join(directory, filename)
  const relativePath = path.relative(publicDir, filePath)

  if (!fs.existsSync(filePath)) {
    transform = '?'
    const entries = Object.entries(params)
    entries.forEach((entry, i) => {
      const [key, value] = entry
      transform = `${transform}${key}=${value}${i < entries.length - 1 ? '&' : ''}`
    })

    await downloadFile(`/assets/${id}${transform}`, filePath)
  }

  return {
    path: relativePath,
    width: params.width 
  }
}
```

If there was an API like the one I mentioned in my previous comment, maybe it should send the raw bytes of the image, as it seems to be better for programmatic use? But I'm not sure :thinking:

**Reply by joselcvarela** [reply]
Hello @ametis70 

I believe the `assets` entity could be tricky for the SDK.
It not feels right to save assets into memory. Instead we should work with streams.
We need to keep in mind should be platform agnostic, i.e., should work both on client and server side. 

Also, your suggestion adds query parameters to `files` entity.
But we should follow the same structure as the API and create a new entity `assets`.

That being said, an example of usage could be:
```ts
import fs from 'fs'

const directus = new Directus();

const readableStream = directus.assets.stream(1, 'public/assets');

readableStream.pipe(fs.createReadStream('my_asset.png'))

**Reply by ametis70** [reply]
I hadn't realized that assets and files are not the same thing (I started using Directus quite recently). I agree with everything you said @joselcvarela. I will rename the discussion to reflect better what I meant.

I will try some things with streams and report here what I find

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
